### ISSUE TITLE
DCR: Strongly Typed Activities

### ISSUE THREAD

**Issue by drub0y** [Microsoft/botbuilder-dotnet#761]
# TL;DR;

Refactor the `Activity` class so it is is no longer a union of all possible activity schemas and instead represents the common properties of all activities and then utilize subclasses for each specific activity type to reduce confusion about which properties have to do with which types of activities as well as being enable to utilize rich language features for determining activity type.

## Overview

Today the `Activity` class is an "uber" class that is a union of all possible properties that appear across all of the activity types. As of this writing, the number of properties stands at at `39`.😱

As a new developer coming into the SDK figuring out what the distinct activity types really are and which properties are actually applicable to which types can be quite daunting.

Today the API design of `Microsoft.Bot.Schema` tries to somewhat solve the problem by defining an interface hierarchy starting with an `IActivity` that represents the base set of properties that are common to all activity types. From there a set of individual interfaces defined that subtype `IActivity` to add more specific properties for a given activity type (e.g. `IMessageActivity`). This was a step in the right direction, but it's the actual implementation that falls short. Instead of a class implementing each distinct interface, the one `Activity` class implements all of these interfaces.

Given this design, a few APIs within the SDK sometimes take/return a  particular interface, but ultimately everything is still just `Activity` under the covers. Some places in the SDK even take an interface and perform upcasting from a particular interface to the full-blown `Activity` class. This means the interfaces are basically a "lie" because developers can't just go implement their own `IActivity` implementations and have them work with the core framework.

## Goals

* Eliminate the "uber" class anti-pattern and replace with a rich, strongly typed hierarchy
* Eliminate the interfaces as they offer no value
* Enable language level type checking/conversions (e.g. `is`, `as` in C#) vs string checking and proprietary "casting" APIs (e.g. `AsXXXActivity()`)

### Activities Should Be Value Objects

First, I would like to posit that activities should really just be [value objects](https://en.wikipedia.org/wiki/Value_object). As such, there is _no need for interfaces_ because they should contain no behavior that should ever vary between implementation. Thus, there is no need for anyone else to implement the interfaces in any different way.

We can completely eliminate all of the interfaces and go with a pure class based hierarchy solution that looks something like this:

```plain
+----------------------------+
|                            |
|         Activity           |
|                            |
+--------+-------------------+
         |
         |   +----------------------------+
         |   |                            |
         +---+     ActivityWithValue      |
         |   |                            |
         |   +--------------+-------------+
         |                  |   +----------------------------+
         |                  |   |                            |
         |                  +---+      MessageActivity       |
         |                  |   |                            |
         |                  |   +----------+-----------------+
         |                  |              |
         |                  |              |   +----------------------------+
         |                  |              |   |                            |
         |                  |              +---+    MessageUpdateActivity   |
         |                  |              |   |                            |
         |                  |              |   +----------------------------+
         |                  |              |
         |                  |              |   +----------------------------+
         |                  |              |   |                            |
         |                  |              +---+ EndOfConversationActivity  |
         |                  |                  |                            |
         |                  |                  +----------------------------+
         |                  |
         |                  |   +----------------------------+
         |                  |   |                            |
         |                  +---+        EventActivity       |
         |                  |   |                            |
         |                  |   +----------------------------+
         |                  |
         |                  |   +----------------------------+
         |                  |   |                            |
         |                  +---+         XXXActivity        |
         |                      |                            |
         |                      +----------------------------+
         |
         |   +----------------------------+
         |   |                            |
         +---+ ConversationUpdateActivity |
         |   |                            |
         |   +----------------------------+
         |
         |   +----------------------------+
         |   |                            |
         +---+         XXXActivity        |
             |                            |
             +----------------------------+

```

> **NOTE:** not all activity types are included in the diagram, only a few to demonstrate some of the proposed hierarchy.

As previously discussed, the `Activity` base class would only contain the minimum set of properties and subclasses enrich from there.

### Abstracting Activity Serialization

IMHO there is already too tight of a coupling to JSON within the SDK and while this DCR doesn't try to solve that problem entirely it does introduce the beginnings of an abstraction over the serialization of activities which could lead to further decoupling from JSON down the line.

The specific problem that _did_ need to be solved with this DCR revolves around the deserialization of the activity JSON schema. Specifically, JSON.NET does not directly support deserilizating this schema to a strongly typed hierarchy. While JSON.NET does have the ability to embed typing with its `__type` annotation, the activity schema does not adhere to that (nor should it IMNSHO). Instead, we need to get "in the middle" of the JSON deserialization process and "sniff" out the `type` of activity, converting that to a runtime type (an `Activity` type) and then, once that type is selected, we can ask JSON.NET to perform the actual deserialization of the JSON document into an instance of that runtime type instance.

As mentioned, the initial support for this abstraction comes in the form of an `IActivitySerializer` interface which looks like this:

#### IActivitySerializer
```CSharp
public interface IActivitySerializer
{
    Task SerializeAsync(Activity activity, Stream stream, CancellationToken cancellationToken = default(CancellationToken));

    Task<Activity> DeserializeAsync(Stream stream, CancellationToken cancellationToken = default(CancellationToken));
}
```

A `JsonActivitySerializer` implementation of this interface then ships "in-the-box" and implements the afforementioned parsing logic utilizing JSON.NET.

Longer term, the `IActivitySerializer` abstraction does pave the way for allowing other serialization formats for activity instances (e.g. `Protobuf`, `MessagePack`, `BSON`, etc).

Finally, the `JsonActivitySerializer` is the default serializer chosen by the integration libraries if no other serializer is provided via a new `BotFrameworkOptions::ActivitySerializer` configuration API.

> NOTE: I would actually propose that ultimately this implementation _not be_ part of Microsoft.Bot.Builder, but instead go into a Microsoft.Bot.Builder.Serialization.Json package. That might be too much to ask today though, especially considering there is _still_ some tight coupling to `JObject` directly in the framework right now.

### Take Advantage of Langauge Features

Today, determing the activity type currently requires string equality checks and then either accessing the properties of the `Activity` that the developer knows apply for that activity or using the `AsXXXActivity()` conversion methods:

#### Old - if statement with string check, then use "known" properties

```CSharp
if (activity.Type == ActivityTypes.Message)
{
    await context.SendAsync("Echo: " + activity.Text);
}
```

#### Old - if statement with string check, then use AsXXXActivity()

```CSharp
if (activity.Type == ActivityTypes.Message)
{
    await context.SendAsync("Echo: " + activity.AsMessageActivity().Text);
}
```

Instead, with a proper strongly typed class hierarchy in place, we can now use the pattern matching features of the language to test/convert the type in one simple, natural way:

#### New - pattern matching if statement

```CSharp
if (activity is MessageActivity message)
{
    await context.SendAsync("Echo: " + message.Text);
}
```

Now, that was the simplest scenario of handling just a single activity type. Consider a more realistic scenario where a bot developer needs to actually handle more than just messages. This quickly gets messier and messier.

They could choose to use stacked if/else checks such as:

#### Old - stacked if/else string checks

```CSharp
if (activity.Type == ActivityTypes.ConversationUpdate)
{
    // ... just use known props or .AsConversationUpdateActivity()
}
else if(activity.Type == ActivityTypes.Message)
{
    // ... just use known props or .AsMessageActivity()
}
else if(activity.Type == ActivityTypes.EndOfConversation)
{
    // ... just use known props or .AsEndOfConversationActivity()
}
```

_Or_ (hopefully) they used a `switch` statement:

#### Old - switch/case on string

```CSharp
switch(activity.Type)
{
    case ActivityTypes.ConversationUpdate:
        // ... just use known props or .AsConversationUpdateActivity()

        break;

    case ActivityTypes.Message:
        // ... just use known props or .AsMessageActivity()

        break;

    case ActivityTypes.EndOfConversation:
        // ... just use known props or .AsEndOfConversationActivity()

        break;
}
```

With strongly typed messages, they can now once again leverage pattern matching in C# 7 to do this far more gracefully :

#### New - using pattern matching switch statement

```CSharp
switch(activity)
{
    case ConversationUpdateActivity conversationUpdate:
        // use strongly typed conversationUpdate variable

        break;

    case MessageActivity message:
        // use strongly typed message variable

        break;

    case EndOfConversationActivity endOfConversation:
        // use strongly typed endOfConversation variable

        break;
}
```

#### Move Behaviors to Extension Methods

There were several methods added both directly the `Activity` class as well as declared on specific `IXXXActivity` interfaces. These included:

* Factory methods for constructing specific activity types
* _N_ `AsXXXActivity` conversion methods 
* Utility methods for working with specific types of activities (e.g. `GetConversationReference`, `ApplyConversationReference`, etc.).

Each of those was dealth with in following ways:

* Factory methods - no longer needed, just construct the `Activity` subclass you actually want and use object initializer syntax to fill out any applicable properties you want
* `AsXXXActivity` conversion methods - are no longer needed due to having actual `Activity` subclasses representing each type
* Utility methods - to keep the `Microsoft.Bot.Schema` assembly as free from behavior as possible, all of the utility methods are moved into extension method classes that are now defined in types in `Microsoft.Bot.Builder`

##### Bonus Side Effect: Delete Test Project for Schema

With the `Activity` type hierarchy as pure value  objects and all behavior removed/migrated, there is no longer any behavior to test in `Microsoft.Bot.Schema` and we can do away with its dedicated test project.

> **NOTE:** Since all behavior was moved to `Microsoft.Bot.Builder`, so were any related tests moved to `Microsoft.Bot.Builder.Tests`.

## Future Considerations

With a property `Activity` type hierarchy, I would consider prototyping a base `Bot` class which implements the standard `IBot` interface, but provides a core implementation of `OnTurn` that actually dispatches specific `Activity` types to specific `OnXXX` virtual methods that can be overriden. This would take the majority of the responsibility of handling dispatching by activity type, one of the most common patterns new developers end up having to figure out pretty early on, out of most people's hands and now they can just override the methods for the types they are interested in.

**Comment by garypretty** [comment]
I love this approach. Especially being able to use the language level checking / conversions.  Plus, anything like this that can make the experience of on boarding new users to the SDK, which I think this absolutely does, is great to see.  I absolutely support this and think it would be great to see this move forward before we get too far down the road to GA.

**Comment by johnataylor** [comment]
The Activity + interfaces model shipped in 4.0. Certainly this proposal has merit, however, this isn't nearly enough of a significantly different model to justify having it as an additional model, so closing this discussion.

However, referring to the "Future Considerations" section that revisits the original "integration" idea of using the object-oriented programming model of abstract classes and inheritance to model the Activity protocol, we now have a new sample that demonstrates just that:

https://github.com/Microsoft/BotBuilder-Samples/tree/master/samples/csharp_dotnetcore/30.asp-mvc-bot
